rfcs/0008-transition-frontier-controller.md

```diff
@@ -0,0 +1,337 @@
+* [Ledger Catchup](#ledger-catchup)
+* [Sync Handler](#sync-handler)
+
+Note that all components communicate between each other solely through the use
+of pipes. By solely using pipes to connect components we can be more explicit about the [Async Control Flow](#async-control-flow).
+
+After describing all the components, we'll cover the specific behavior of
+the [Async Control Flow](#async-control-flow) holistically.
+
+Before covering any of the components it's worth going over some fundamental
+changes in behavior and structure in more detail in [big Changes](#big-changes)
+
+<a href="big-changes"></a>
+### Big Changes
+
+Major differences in behavior between this component and the existing ledger-
+builder-controller are as follows:
+
+1. We only sync-ledger to the locked-ledger (if necessary) and do a new
+[History sync](#history-sync) for catchup jobs
+
+Rationale: Without history sync, we are vulnerable to believing an invalid
+state: Certain parts of the ledger-builder state are not checked in a SNARK and
+so we must get enough info to go back to the locked state (where we know we've
+achieved consensus).
+
+2. We take advantage of the merkle-mask to instead of transitions, putting a `Breadcrumb.t` which contains an `External_transition.t` and a now light-weight `Ledger_builder.t` it also has a notion of the prior breadcrumb. See [RFC-0007](0007-persistent-ledger-builder-controller) for more info on merkle masks, and [Transition frontier](#transition-frontier) for more info about how these masks are configured.
+
+Rationale: We no longer need the asynchronous `Path_traversal` job that is present in the current ledger-builder-controller (simplifying complexity), and the frequent operation of answering sync-ledger queries no longer require materializing ledger-builders.
+
+3. The `Ktree.t` is mutable and exposes an $O(1)$ `lookup : State_hash.t -> node_entry` where `node_entry` contains (in our case) a `Breadcrumb.t`. We will discuss this further in [Ktree](#ktree).
+
+Rationale: We perform `lookup` and `add` frequently on this structure and `add` can be $O(1)$ in the presence of this `lookup` too. `lookup` also lets us traverse the tree forwards and backwards without explicit backedges in our [rose tree](https://en.wikipedia.org/wiki/Rose_tree) -- further simplifying the `Ktree.t` implementation.
+
+<a href="transition-handler"></a>
+### Transition Handler
+
+The transition handler is broken down into two main components: Validator and
+Processor
+
+#### Validator
+
+* Input: `External_transition.t`
+* Output: `External_transition.t` (we should consider making an `External_transition.Validated.t` for output here)
+
+The validator receives as input external transitions from the network and
+performs a series of checks on the transition before passing it off to be
+processed. For simplicity it can also receive transitions created by the local
+proposer. In the future, we can consider skipping this step.
```

Comment:
I have skipped this step for now. I couldn't think of how we should be handling them differently. I think maybe the only thing we should consider doing is creating a priority pipe that will push any of our own transitions to the front of the pipe or something.